### ISSUE TITLE
Erroneous deadlock when 2 threads wait on each other on MVars

### ISSUE THREAD

**Issue by Mikolaj** [ghcjs/ghcjs#454]
If needed, I will come up with a small example. In the actual code, one thread waits here

https://github.com/LambdaHack/LambdaHack/blob/6f9eca34c59c500057a581b3b8ed561a3fd8a1e6/Game/LambdaHack/Client/UI/Frontend.hs#L103

and another here:

https://github.com/LambdaHack/LambdaHack/blob/6f9eca34c59c500057a581b3b8ed561a3fd8a1e6/Game/LambdaHack/Client/UI/Frontend.hs#L124

With standard GHC, there is no deadlock, because neither MVar is taken anywhere else in the code, and for each taking of any of the two MVars, the other one is being put.

The deadlock occurs after the `fshowNow` MVar is taken twice (and so the `ftimeout` MVar is taken at least once). 

Edit: GHCJS version is today's Quick Start snapshot, GHC is 7.10.3.

Edit2: the CPU is almost idle when the deadlock happens.

Edit3: when I eliminate one of the MVars and wait on the other one both for taking and putting, the code works fine: https://github.com/LambdaHack/LambdaHack/blob/248e29aa0b2c2cec782beb4db6859193c6ccf041/Game/LambdaHack/Client/UI/Frontend.hs#L120

**Comment by luite** [comment]
Can you make a small example that reproduces this? Ideally something that just prints some things to the terminal when run with node.js (running the same program with GHC should give the expected result).

I'd like to fix this and add a test to the test suite.

Thanks!

**Comment by Mikolaj** [comment]
I can't find the time right now, but I've created a branch that exhibits the problem: https://github.com/LambdaHack/LambdaHack/tree/ghcjs_mvar_problem and tested with node.js. Feel free to minimize the example.

When you configure with `cabal configure --ghcjs`, build and run with `node dist/build/LambdaHack/LambdaHack.jsexe/all.js --frontendStd`, it hangs after a few screenfuls.

When you configure with `cabal configure` or `cabal configure -fvty` (the latter does not require gtk2hs), build and run with `dist/build/LambdaHack/LambdaHack --frontendStd`, it runs forever, outputing lots of text.

As I explained earlier, replacing two mvars with one mvar fixes the problem, hence my conjecture it's caused by a different behaviour of concurrency in ghcjs, or at least related to it.
